node/src/node/slave/state/preparation.cpp

```diff
@@ -33,30 +33,39 @@ namespace ph = std::placeholders;
 using asio::ip::tcp;
 
 preparation_t::preparation_t(std::shared_ptr<machine_t> slave_) :
-    slave(std::move(slave_))
+    slave(std::move(slave_)),
+    is_terminated(false)
 {}
 
 auto preparation_t::name() const noexcept -> const char* {
     return "preparation";
 }
 
 auto preparation_t::terminate(const std::error_code& ec) -> void {
-    slave->shutdown(ec);
+    is_terminated.apply([&] (bool& is_terminated) {
+        slave->shutdown(ec);
```

Comment:
Should we check here that `is_terminated` has already been set and skip calling `slave->shutdown()` in this case?

Reply:
Yep, it makes sense!